src/types/YText.js

```diff
@@ -164,6 +164,20 @@ const insertNegatedAttributes = (transaction, parent, currPos, negatedAttributes
   const doc = transaction.doc
   const ownClientId = doc.clientID
   negatedAttributes.forEach((val, key) => {
+    // check if we really need to create attributes
+    //  (the attribute may be set the desired value already)
+    let n = currPos.right
+    while(
```

Comment:
I think the problem is that we want to insert a negated attribute, but the negated attribute might already exist in one of the following `ContentFormat` items. So, when we insert the same `ContentFormat` twice, we end up with the undesired result.

The ideal solution, it seems to me is to avoid inserting the negated attribute if it is already inserted in one of the following `ContentFormat` items.

This approach, however, seems to iterate through the whole list of `ContentFormat` for each entry in `negatedAttributes` (O(N•M)). Then it deletes the entries that we want to insert anyway.

I suggest that you simply `return` when you find an item that was already inserted with the same key (the values shouldn't matter here). There is no need to delete it and `return` (we would end up without negatedAttributes if we deleted an item, but don't insert anything).

Reply:
This makes sense, i changed the pull request to use the previous cleanup loop, this will now also check the attributes after the non-format item in question.

This does basically the same (delete and skip the negated attribute) but without O(N•M) 😅 

A quick step-by-step:

This happens with the current yjs release version (yjs@13.5.27):
```js
  >  initial
   string ID { client: 574277975, clock: 0 } ContentString { str: 'Attack ships ' }
   format ID { client: 574277975, clock: 47 } ContentFormat { key: 'bold', value: true }
   string ID { client: 574277975, clock: 13 } ContentString { str: 'on fire' }
   format ID { client: 574277975, clock: 48 } ContentFormat { key: 'bold', value: null }
   string ID { client: 574277975, clock: 20 } ContentString { str: ' off the shoulder of Orion.' }
  >  after findPosition
   string ID { client: 574277975, clock: 0 } ContentString { str: 'Attack ships ' }
   format ID { client: 574277975, clock: 47 } ContentFormat { key: 'bold', value: true }
   string ID { client: 574277975, clock: 13 } ContentString { str: 'on ' }
   string ID { client: 574277975, clock: 16 } ContentString { str: 'fire' }
   format ID { client: 574277975, clock: 48 } ContentFormat { key: 'bold', value: null }
   string ID { client: 574277975, clock: 20 } ContentString { str: ' off the shoulder of Orion.' }
  >  after minimizeAttributeChanges
   string ID { client: 574277975, clock: 0 } ContentString { str: 'Attack ships ' }
   format ID { client: 574277975, clock: 47 } ContentFormat { key: 'bold', value: true }
   string ID { client: 574277975, clock: 13 } ContentString { str: 'on ' }
   string ID { client: 574277975, clock: 16 } ContentString { str: 'fire' }
   format ID { client: 574277975, clock: 48 } ContentFormat { key: 'bold', value: null }
   string ID { client: 574277975, clock: 20 } ContentString { str: ' off the shoulder of Orion.' }
  >  after insertAttributes
   string ID { client: 574277975, clock: 0 } ContentString { str: 'Attack ships ' }
   format ID { client: 574277975, clock: 47 } ContentFormat { key: 'bold', value: true }
   string ID { client: 574277975, clock: 13 } ContentString { str: 'on ' }
   format ID { client: 574277975, clock: 49 } ContentFormat { key: 'bold', value: null }
   string ID { client: 574277975, clock: 16 } ContentString { str: 'fire' }
   format ID { client: 574277975, clock: 48 } ContentFormat { key: 'bold', value: null }
   string ID { client: 574277975, clock: 20 } ContentString { str: ' off the shoulder of Orion.' }
  >  after delete loop and quill check
   string ID { client: 574277975, clock: 0 } ContentString { str: 'Attack ships ' }
   format ID { client: 574277975, clock: 47 } ContentFormat { key: 'bold', value: true }
   string ID { client: 574277975, clock: 13 } ContentString { str: 'on ' }
   format ID { client: 574277975, clock: 49 } ContentFormat { key: 'bold', value: null }
   string ID { client: 574277975, clock: 16 } ContentString { str: 'fire' }
   format ID { client: 574277975, clock: 48 } ContentFormat { key: 'bold', value: null }
   string ID { client: 574277975, clock: 20 } ContentString { str: ' off the shoulder of Orion.' }
  >  after insertNegatedAttributes
   string ID { client: 574277975, clock: 0 } ContentString { str: 'Attack ships ' }
   format ID { client: 574277975, clock: 47 } ContentFormat { key: 'bold', value: true }
   string ID { client: 574277975, clock: 13 } ContentString { str: 'on ' }
   format ID { client: 574277975, clock: 49 } ContentFormat { key: 'bold', value: null }
   string ID { client: 574277975, clock: 16 } ContentString { str: 'fire' }
   format ID { client: 574277975, clock: 50 } ContentFormat { key: 'bold', value: true }
   format ID { client: 574277975, clock: 48 } ContentFormat { key: 'bold', value: null }
   string ID { client: 574277975, clock: 20 } ContentString { str: ' off the shoulder of Orion.' }
```

When we just skip the negated attribute (with return) we still get the `{bold: null}` twice:
```js
  >  initial
   string ID { client: 1843338659, clock: 0 } ContentString { str: 'Attack ships ' }
   format ID { client: 1843338659, clock: 47 } ContentFormat { key: 'bold', value: true }
   string ID { client: 1843338659, clock: 13 } ContentString { str: 'on fire' }
   format ID { client: 1843338659, clock: 48 } ContentFormat { key: 'bold', value: null }
   string ID { client: 1843338659, clock: 20 } ContentString { str: ' off the shoulder of Orion.' }
  >  after findPosition
   string ID { client: 1843338659, clock: 0 } ContentString { str: 'Attack ships ' }
   format ID { client: 1843338659, clock: 47 } ContentFormat { key: 'bold', value: true }
   string ID { client: 1843338659, clock: 13 } ContentString { str: 'on ' }
   string ID { client: 1843338659, clock: 16 } ContentString { str: 'fire' }
   format ID { client: 1843338659, clock: 48 } ContentFormat { key: 'bold', value: null }
   string ID { client: 1843338659, clock: 20 } ContentString { str: ' off the shoulder of Orion.' }
  >  after minimizeAttributeChanges
   string ID { client: 1843338659, clock: 0 } ContentString { str: 'Attack ships ' }
   format ID { client: 1843338659, clock: 47 } ContentFormat { key: 'bold', value: true }
   string ID { client: 1843338659, clock: 13 } ContentString { str: 'on ' }
   string ID { client: 1843338659, clock: 16 } ContentString { str: 'fire' }
   format ID { client: 1843338659, clock: 48 } ContentFormat { key: 'bold', value: null }
   string ID { client: 1843338659, clock: 20 } ContentString { str: ' off the shoulder of Orion.' }
  >  after insertAttributes
   string ID { client: 1843338659, clock: 0 } ContentString { str: 'Attack ships ' }
   format ID { client: 1843338659, clock: 47 } ContentFormat { key: 'bold', value: true }
   string ID { client: 1843338659, clock: 13 } ContentString { str: 'on ' }
   format ID { client: 1843338659, clock: 49 } ContentFormat { key: 'bold', value: null }
   string ID { client: 1843338659, clock: 16 } ContentString { str: 'fire' }
   format ID { client: 1843338659, clock: 48 } ContentFormat { key: 'bold', value: null }
   string ID { client: 1843338659, clock: 20 } ContentString { str: ' off the shoulder of Orion.' }
  >  after delete loop and quill check
   string ID { client: 1843338659, clock: 0 } ContentString { str: 'Attack ships ' }
   format ID { client: 1843338659, clock: 47 } ContentFormat { key: 'bold', value: true }
   string ID { client: 1843338659, clock: 13 } ContentString { str: 'on ' }
   format ID { client: 1843338659, clock: 49 } ContentFormat { key: 'bold', value: null }
   string ID { client: 1843338659, clock: 16 } ContentString { str: 'fire' }
   format ID { client: 1843338659, clock: 48 } ContentFormat { key: 'bold', value: null }
   string ID { client: 1843338659, clock: 20 } ContentString { str: ' off the shoulder of Orion.' }
  >  after insertNegatedAttributes
   string ID { client: 1843338659, clock: 0 } ContentString { str: 'Attack ships ' }
   format ID { client: 1843338659, clock: 47 } ContentFormat { key: 'bold', value: true }
   string ID { client: 1843338659, clock: 13 } ContentString { str: 'on ' }
   format ID { client: 1843338659, clock: 49 } ContentFormat { key: 'bold', value: null }
   string ID { client: 1843338659, clock: 16 } ContentString { str: 'fire' }
   string ID { client: 1843338659, clock: 20 } ContentString { str: ' off the shoulder of Orion.' }
```

With the updated loop (same result my previous solution):
```js
  >  initial
   string ID { client: 1019279783, clock: 0 } ContentString { str: 'Attack ships ' }
   format ID { client: 1019279783, clock: 47 } ContentFormat { key: 'bold', value: true }
   string ID { client: 1019279783, clock: 13 } ContentString { str: 'on fire' }
   format ID { client: 1019279783, clock: 48 } ContentFormat { key: 'bold', value: null }
   string ID { client: 1019279783, clock: 20 } ContentString { str: ' off the shoulder of Orion.' }
  >  after findPosition
   string ID { client: 1019279783, clock: 0 } ContentString { str: 'Attack ships ' }
   format ID { client: 1019279783, clock: 47 } ContentFormat { key: 'bold', value: true }
   string ID { client: 1019279783, clock: 13 } ContentString { str: 'on ' }
   string ID { client: 1019279783, clock: 16 } ContentString { str: 'fire' }
   format ID { client: 1019279783, clock: 48 } ContentFormat { key: 'bold', value: null }
   string ID { client: 1019279783, clock: 20 } ContentString { str: ' off the shoulder of Orion.' }
  >  after minimizeAttributeChanges
   string ID { client: 1019279783, clock: 0 } ContentString { str: 'Attack ships ' }
   format ID { client: 1019279783, clock: 47 } ContentFormat { key: 'bold', value: true }
   string ID { client: 1019279783, clock: 13 } ContentString { str: 'on ' }
   string ID { client: 1019279783, clock: 16 } ContentString { str: 'fire' }
   format ID { client: 1019279783, clock: 48 } ContentFormat { key: 'bold', value: null }
   string ID { client: 1019279783, clock: 20 } ContentString { str: ' off the shoulder of Orion.' }
  >  after insertAttributes
   string ID { client: 1019279783, clock: 0 } ContentString { str: 'Attack ships ' }
   format ID { client: 1019279783, clock: 47 } ContentFormat { key: 'bold', value: true }
   string ID { client: 1019279783, clock: 13 } ContentString { str: 'on ' }
   format ID { client: 1019279783, clock: 49 } ContentFormat { key: 'bold', value: null }
   string ID { client: 1019279783, clock: 16 } ContentString { str: 'fire' }
   format ID { client: 1019279783, clock: 48 } ContentFormat { key: 'bold', value: null }
   string ID { client: 1019279783, clock: 20 } ContentString { str: ' off the shoulder of Orion.' }
  >  after delete loop and quill check
   string ID { client: 1019279783, clock: 0 } ContentString { str: 'Attack ships ' }
   format ID { client: 1019279783, clock: 47 } ContentFormat { key: 'bold', value: true }
   string ID { client: 1019279783, clock: 13 } ContentString { str: 'on ' }
   format ID { client: 1019279783, clock: 49 } ContentFormat { key: 'bold', value: null }
   string ID { client: 1019279783, clock: 16 } ContentString { str: 'fire' }
   string ID { client: 1019279783, clock: 20 } ContentString { str: ' off the shoulder of Orion.' }
  >  after insertNegatedAttributes
   string ID { client: 1019279783, clock: 0 } ContentString { str: 'Attack ships ' }
   format ID { client: 1019279783, clock: 47 } ContentFormat { key: 'bold', value: true }
   string ID { client: 1019279783, clock: 13 } ContentString { str: 'on ' }
   format ID { client: 1019279783, clock: 49 } ContentFormat { key: 'bold', value: null }
   string ID { client: 1019279783, clock: 16 } ContentString { str: 'fire' }
   string ID { client: 1019279783, clock: 20 } ContentString { str: ' off the shoulder of Orion.' }
```